rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -968,326 +1023,193 @@ object ExecutionBatch extends Loggable {
+                case Some(v) => //found a new pair!
+                  recPairReports(tail, newFreeValues, v :: newPairedValues, unexpected, mode)
+              }
+          }
+      }
     }
 
-    /*
-     * And now, merge all values into a component.
-     */
-    ComponentStatusReport(
-        expectedComponent.componentName
-      , ComponentValueStatusReport.merge(unexpectedStatusReports ++ noneValue ++ simpleValues ++ cfeVarValues ++ lastUnexpected)
-    )
-  }
-
-
-  /*
-   * Recursively look into remaining reports for CFEngine variables,
-   * values are accumulated and we remove reports while we use them.
-   *
-   * The first matching pattern is used as the correct one, so
-   * they must be sorted from the most specific to the less one
-   * if you don't want to see things like in #7758 happen
-   *
-   * The returned reports are the one that are not matches by
-   * any cfengine variable value
-   *
-   * Visibility is for test
-   *
-   */
-  private[reports] def extractCFVarsFromReports (
-      cfengineVars: List[(String, Pattern)]
-    , allReports  : List[Reports]
-    , noAnswerType: ReportType
-    , policyMode  : PolicyMode
-  ) : (Seq[ComponentValueStatusReport], Seq[Reports]) = {
 
-    /*
-     * So, we don't have any simple, robust way to sort
-     * reports or patterns to be sure that we are not
-     * introducing loads of error case.
-     *
-     * So, we are going to test all patterns on all regex,
-     * and find the combination with the most report used.
-     *
-     * One optimisation thought: when a pattern match exactly
-     * one report, we can prune that couple for remaining
-     * pattern tests.
-     *
-     * It is known that that way of testing is quadratic and
-     * will takes A LOT OF TIME for big input.
-     * The rationnal to do it none the less is:
-     * - the quadratic nature is only value with CFEngine params,
-     * - the time remains ok below 20 or so entries,
-     * - the pruning helps
-     * - there is a long terme solution with the unique
-     *   identification of a report for a component (and so no more
-     *   guessing)
-     *
-     * Given all that, if an user fall in the quadratic nature, we
-     * can workaround the problem by splitting his long list of
-     * problematic patterns into two directives, with a little
-     * warn log message.
-     */
+    val componentGotAtLeastOneReport = filteredReports.nonEmpty
 
+    // the list of expected (value, unexpanded value for display)
+    // it is very important to sort pattern so that the more precise come first to avoid
+    // bugs like #7758. A pattern specificity, in our case, can somehow be told from
+    // the lenght of the pattern when \Q\E.* are removed.
+    //
+    val values = expectedComponent.groupedComponentValues.toList.map { case(v, u) =>
+      val isVar = matchCFEngineVars.pattern.matcher(v).matches()
+      val pattern = replaceCFEngineVars(v)
```

Comment:
shouldn't we evaluate the pattern only if "isVar" is true ?

Reply:
no, we need a pattern for all component values, to be able to only works on pattern matching in `findMatchingValue()`

Reply:
isn't it super expensive ? Plus from the doc of replaceCFEngineVars, it should be called only for string containing a cfengine var

Reply:
ok -thank you for the explanation.